### PR TITLE
[Debuginfo][TailCallElim] Fix #86262: drop the debug location of entry branch

### LINKED ISSUE #86262: [Debuginfo][TailCallElim] Misleading stepping caused by new BranchInst in NewEntry
The source code:
```C
int func(int a) {
    if (a > 10) {
        return 5 * func(a - 1);
    }
    if (a > 1) {
        return a * func(a - 1);
    } 
    return 1;
}

int main() {
    func(1);
}
```

The compilation commands:
```bash
$ clang -S -emit-llvm -Xclang -disable-O0-optnone main.c -o main.ll -g
$ opt -S -passes=mem2reg,tailcallelim main.ll -o opt.ll
$ clang opt.ll
```

The misleading stepping:
```bash
$ debug a.out 
(lldb) target create "a.out"
Current executable set to '/llvm-test/TailCallElim/a.out' (x86_64).
(lldb) b func
Breakpoint 1: where = a.out`func + 8 at main.c:3:20, address = 0x0000000000001138
(lldb) r
Process 2859495 launched: '/llvm-test/TailCallElim/a.out' (x86_64)
Process 2859495 stopped
* thread #1, name = 'a.out', stop reason = breakpoint 1.1
    frame #0: 0x0000555555555138 a.out`func(a=<unavailable>) at main.c:3:20
   1    int func(int a) {
   2        if (a > 10) {
-> 3            return 5 * func(a - 1);
   4        }
   5        if (a > 1) {
   6            return a * func(a - 1);
   7        } 
(lldb) s
Process 2859495 stopped
* thread #1, name = 'a.out', stop reason = step in
    frame #0: 0x0000555555555154 a.out`func(a=1) at main.c:2:11
   1    int func(int a) {
-> 2        if (a > 10) {
   3            return 5 * func(a - 1);
   4        }
   5        if (a > 1) {
   6            return a * func(a - 1);
   7        } 
(lldb) v
(int) a = 1
```

The debugger gets into then branch when the condition is not satisfied. This is caused by [the newly created BranchInst `BI` in `NewEntry`](https://github.com/llvm/llvm-project/blob/2fe81edef6f0b35abffbbc59b649b30ea9c15a62/llvm/lib/Transforms/Scalar/TailRecursionElimination.cpp#L513). `BI` is inserted in the entry of the function (before the first if-statement), however, it preserves the debug location of the tail recursion call `CI` (the call `func(a-1)`), which could be located in a conditional branch.

The debug location of `BI` should be dropped.

### PR DESCRIPTION
**Pull request by Apochens**
This pr fixes #86262.

### CODE REVIEW

**Comment by Apochens** [comment]
@jmorse @SLTozer @dwblaikie Could you please review this?

**Comment by SLTozer** [comment]
For requesting reviews, you should add the prospective reviewers to the "Reviewers" list in the PR; @ mentions are better for one-off questions where you need a specific person's reply, and won't add the issue to their "Review requested" list.

**Comment by Apochens** [comment]
Yeah, I agree that. But I don't have the permission to add reviewers in the PR.😞

**Review by SLTozer — COMMENTED** [review]
Largely LGTM, one inline comment.

File: llvm/lib/Transforms/Scalar/TailRecursionElimination.cpp
Comment (Apochens) [reply llvm/lib/Transforms/Scalar/TailRecursionElimination.cpp]: Ok, that looks more suitable.

File: llvm/test/Transforms/TailCallElim/entry-branch-drop-debugloc.ll
Comment (SLTozer) [inline llvm/test/Transforms/TailCallElim/entry-branch-drop-debugloc.ll:4]:
```suggestion
; CHECK: entry:
```

File: llvm/test/Transforms/TailCallElim/entry-branch-drop-debugloc.ll
Comment (SLTozer) [inline llvm/test/Transforms/TailCallElim/entry-branch-drop-debugloc.ll:1]: One more thing, it'd be good to add a comment describing what this is testing for, particularly since the CHECK lines are fairly bare. Related to that, does this test add any coverage that the previous test (`debugloc.ll`) doesn't cover?

**Review by SLTozer — COMMENTED** [review]
A couple more inline comments about the test - if it turns out the test isn't needed it can be deleted, but if it's adding unique coverage then LGTM.

File: llvm/test/Transforms/TailCallElim/entry-branch-drop-debugloc.ll
Comment (Apochens) [reply llvm/test/Transforms/TailCallElim/entry-branch-drop-debugloc.ll:1]: This test case indeed adds coverage (related to the creation of the recursion accumulator). However, the purpose of the tests is to check that the new entry branch has no debugloc. For this, the previous test is enough. I'll delete the newly added test.

**Review by SLTozer — APPROVED** [review]

**Comment by Apochens** [comment]
Hi! I would like to politely ask if this PR is ready to merge?

**Comment by SLTozer** [comment]
> Hi! I would like to politely ask if this PR is ready to merge?

Yep, you have the green tick of approval!

File: llvm/lib/Transforms/Scalar/TailRecursionElimination.cpp
Comment (mikaelholmen) [inline llvm/lib/Transforms/Scalar/TailRecursionElimination.cpp:512]: BI is unused now (but referenced in the comment at line 515), leading to compiler warnings.
```
../lib/Transforms/Scalar/TailRecursionElimination.cpp:512:15: error: unused variable 'BI' [-Werror,-Wunused-variable]
  BranchInst *BI = BranchInst::Create(HeaderBB, NewEntry);
              ^
1 error generated.
```

Comment (Apochens) [reply llvm/lib/Transforms/Scalar/TailRecursionElimination.cpp:512]: Thanks for pointing out this. I fix this in #86927.

**Comment by jryans** [comment]
> But it seems that this PR has remained unmerged for a while 🤔

LLVM PRs tend to wait a bit (sometimes a few days) even after the first approving review in case others may have feedback to add. There's no firm rule though. With simpler changes, it's typically fine to merge soon after the first approval.

Sometimes reviewers forget that new contributors (such as yourself) can't merge the PR themselves (when the PR author has commit access, the typical process is the PR author merges once there's approval). You may need to comment on PRs after approval to remind reviewers that you don't have commit access yourself, so they should merge for you.

**Comment by jryans** [comment]
Thanks for working on this! 😄

**Pull request merged**
